src/mds/Server.cc

```diff
@@ -180,6 +213,69 @@ Server::Server(MDSRank *m) :
   supported_features = feature_bitset_t(CEPHFS_FEATURES_MDS_SUPPORTED);
 }
 
+bool Server::propagate_subtree_roots_rstat(MDRequestRef& mdr, CInode* cur)
+{
+  CDir* pdir = cur->get_projected_parent_dn()->get_dir();
+  set<CDir*> dirs;
+  mdcache->get_wouldbe_subtree_bounds(pdir, dirs);
+  list<CDir*> drls(dirs.begin(), dirs.end());
+
+  bool went_wait = false;
+  
+  for (auto dir : drls) {
+    if (!cur->is_projected_ancestor_of(dir->get_inode()))
+      continue;
+    
+    dout(20) << "found a subtree: " << *dir << dendl;
+    
+    if (dir->is_ambiguous_auth()) {
+      dir->add_waiter(CDir::WAIT_SINGLEAUTH, new C_MDS_RetryRequest(mdcache, mdr));
+      dout(20) << "dir " << *dir << "is ambiguous auth, going to wait for single auth" << dendl;
+      went_wait = true;
+      continue;
+    }
+    
+    if (dir->is_auth()) {
+      dout(20) << "dir " << *dir << "is auth, skipping" << dendl;
+      dirs.clear();
+      mdcache->get_wouldbe_subtree_bounds(dir, dirs);
+      list<CDir*> tmpls(dirs.begin(), dirs.end());
+      drls.splice(drls.end(), tmpls, tmpls.begin(), tmpls.end());
+      continue;
+    }
+    
+    if (mdr->more()->queried_for_rstat_propagation.count(dir)) {
+      dout(20) << "dir " << *dir << "has been queried, skipping" << dendl;
+      continue;
+    }
+
+    if (!dir->get_inode()->is_auth()) {
+      dout(20) << " I'm not auth for " << *dir->get_inode() << ", skipping" << dendl;
+      continue;
+    }
+    
+    for (auto &p : dir->get_inode()->get_replicas()) {
+      if (p.first != dir->get_dir_auth().first)
+	continue;
+      dout(20) << "found auth for " << *dir << ", quering" << dendl;
+      MDRequestRef internal_mdr = mds->mdcache->request_start_internal(CEPH_MDS_OP_PROPAGATE_RSTAT);
+      internal_mdr->parent_mdr = mdr;
+      MMDSSlaveRequest* req = new MMDSSlaveRequest(internal_mdr->reqid, internal_mdr->attempt,
+	  MMDSSlaveRequest::OP_PROPAGATERSTATS);
```

Comment:
why starting new request.

Reply:
@ukernel The reason that I added a new request is that: I can't guarantee that all dirtied locks are nudged in one run, some of them may have to some "writebehind" or "scatter/gather" procedures, I need a way to retry the nudging when those procedures complete, so I added a new request and accomplish the retrying of the nudging by the "C_MDS_RetryRequest" callback

Reply:
@ukernel Hi, sorry, I misunderstood you. The reason that I added this new request is that, if I use the slave request directly, all the slave requests would have the same reqid, and that would trigger some assertions(Sorrry that I don't remenber which specific assertion) when dealing with slave request replies.

Reply:
@ukernel Hi, sorry again, I now recall that why I added this new request. Consider the following scenario: subtree /a/b is on MDS x, subtree /a/b/c/d is on MDS y, subtree /a/b/c/d/e/f is on MDS x. If I use the slave request here directly, when MDS send out a slave request to MDS x for propagating subtree /a/b/c/d/e/f, that slave request would have the same reqid as the one MDS x sent to MDS y for propagating /a/b/c/d, which would trigger the assertion "assert(mdr->slave_request == 0)" in "handle_slave_request". So I added a new request, constructing a slave request hierarchy.